### PR TITLE
Debian installers

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/aspnet/MetaPackages/issues/149

### CODE REVIEW

File: tools/debian_config.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{
+    "maintainer_name":"Microsoft",
```
Comment (JunTaoLuo) [inline tools/debian_config.json]: @Eilon Please review this debian config file. This is mostly copied over from https://github.com/dotnet/cli/blob/3419a87d6fb620a2ea7981cdd6cd43db3b17cb5f/packaging/deb/dotnet-debian_config.json

Nodes especially important are maintainer_*, maintainer_email, package_name, install_root, *_description, homepage, copyright and license. The values in release and control seem to be dummy values though I need to verify them.

**Comment by JunTaoLuo** [comment]
The generation of deb installers is ready. I'll split the rpm generation into a different PR since that uses a different tool and requires some more work.

File: build/repo.targets [hunk 2]
```diff
@@ -289,4 +301,136 @@
...
+
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=debian" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.14.04" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.16.04" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.16.10" />
+  </Target>
+
+  <Target Name="BuildRuntimeStoreDeb">
+    <PropertyGroup>
+      <RuntimeStoreTimestampDebianPackageName>aspnetcore-store-$(TimestampVersion)</RuntimeStoreTimestampDebianPackageName>
```
Comment (JunTaoLuo) [inline build/repo.targets]: This follows the CLI naming scheme:

Package name: aspnetcore-store-{version}, in the CLI: https://github.com/dotnet/cli/blob/358568b07f16749108dd33e7fea2f2c84ccf4563/build/package/Installer.DEB.targets#L30
Package file name: aspnetcore-store-{version}-{RID}, in the CLI: https://github.com/dotnet/cli/blob/358568b07f16749108dd33e7fea2f2c84ccf4563/build/Branding.props#L31

File: build/repo.targets [hunk 3]
```diff
@@ -289,4 +301,136 @@
...
+
+    <!-- Clear the output dir -->
+    <RemoveDir Directories="$(RuntimeStoreInstallerOutputTimeStampDirectory)" />
+    <RemoveDir Directories="$(RuntimeStoreInstallerOutputNoTimeStampDirectory)" />
+    <MakeDir Directories="$(RuntimeStoreInstallerOutputTimeStampDirectory)" />
+    <MakeDir Directories="$(RuntimeStoreInstallerOutputNoTimeStampDirectory)" />
+
+    <!-- Build Docker Image -->
+    <Exec
+      Command="docker build --build-arg USER_ID=%24(id -u) -t docker-image-$(InstallerPlatform) $(InstallerPlatform)"
```
Comment (natemcmaster) [inline build/repo.targets]: Would be good to assert that the docker host is using Linux containers, not windows containers.
```
$ docker version -f "{{ .Server.Os }}"
linux
```

File: build/repo.targets [hunk 4]
```diff
@@ -289,4 +301,136 @@
...
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=debian" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.14.04" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.16.04" />
+    <MSBuild Projects="$(ProjectPath)" Targets="BuildRuntimeStoreDeb" Properties="InstallerPlatform=ubuntu.16.10" />
+  </Target>
+
+  <Target Name="BuildRuntimeStoreDeb">
+    <PropertyGroup>
+      <RuntimeStoreTimestampDebianPackageName>aspnetcore-store-$(TimestampVersion)</RuntimeStoreTimestampDebianPackageName>
+      <RuntimeStoreNoTimestampDebianPackageName>aspnetcore-store-$(NoTimestampVersion)</RuntimeStoreNoTimestampDebianPackageName>
```
Comment (natemcmaster) [inline build/repo.targets]: Have we verified the naming of runtimes-store installers with the PMs? //cc @richlander, cref https://github.com/dotnet/designs/issues/2

**Comment by JunTaoLuo** [comment]
@natemcmaster This actually requires an additional configuation on TeamCity that runs after coherence-signed on a Linux machine. We don't have a coherence-final configuration so I'm just using MetaPackages.

**Comment by natemcmaster** [comment]
We can make new configurations if required. It just seems pointless to have this code in aspnet/Metapackages if it depends on the way coherence-signed works as Coherence-signed is private. It would be better to ensure this code works on its own.

Context: there is a "build from source" effort currently underway to make it easier for people to build asp/.net core on their own.

File: build/repo.targets (hunk 4 above)
Comment (JunTaoLuo) [inline build/repo.targets]: I'm following the convention highlighted by @DamianEdwards in https://github.com/aspnet/MetaPackages/issues/124 though I'm not sure if this is correct. That's why I called out on it in my previous comment. This needs to be checked by a PM.

File: build/repo.targets (hunk 3 above)
Comment (JunTaoLuo) [inline build/repo.targets]: Assert added.

File: tools/docker/debian/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#
+# Copyright (c) .NET Foundation and contributors. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (Eilon) [inline tools/docker/debian/Dockerfile]: Update these to Apache.

**Comment by JunTaoLuo** [comment]
Tested installation and uninstallation using the 16.04 .deb installer. When installed via dpkg, the store will be listed as follows (with other dotnet components as reference):
image

File: tools/RestoreRuntimeStore/RestoreRuntimeStore.csproj [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <Import Project="..\..\build\dependencies.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <DotNetCliToolReference Include="Build.RS" Version="$(AspNetCoreVersion)" />
```
Comment (natemcmaster) [inline tools/RestoreRuntimeStore/RestoreRuntimeStore.csproj]: Did you mean `PackageReference`?

Comment (JunTaoLuo) [inline tools/RestoreRuntimeStore/RestoreRuntimeStore.csproj]: copy pasta

**Comment by JunTaoLuo** [comment]
Got an offline :shipit: from @glennc for the installer names. Will merge soon.

**Pull request merged**